### ISSUE TITLE
MountPropagation breaks Kubernetes on macOS

### ISSUE THREAD

**Issue by jsafrane** [kubernetes/kubernetes#62397]
/kind bug

There is no way you can run the mount --make-rshared / in Docker VM and the shared folders use fuse.osxfs and are dynamically mounted into containers.

Adventurous macOS users must run Kubernetes with `MountPropagation=false` flag. This is OK for now, but the flag will be removed when the feature reaches GA.

There are few options we can do:

* Detect `darwin` in kubelet and turn off the feature silently, even after removal of MountPropagation feature.
* Rework the feature to default to `private` mount propagation, either everywhere or on `darwin`. All pods would run until user runs a pod with explicitly enabled slave or shared mount propagation and he/she would get clear error in this case.

/sig storage
/sig node

**Comment by liggitt** [comment]
> Detect darwin in kubelet

is that even detectable when running the kubelet in a container? I would have expected the OS to be linux

**Comment by jsafrane** [comment]
> is that even detectable when running the kubelet in a container? I would have expected the OS to be linux

That's right. I don't have macOS, bu @mfojtik confirmed it runs in a Linux VM there. So no detection in runtime.

**Comment by thockin** [comment]
One more reason to  abandon support for running tests on MacOS. :)

**Comment by mfojtik** [comment]
@jsafrane you can't detect darwin because docker and kubelet runs in Linux VM, so for kubelet it will look like it runs on Linux. The problem is how Docker for Mac is designed and the way how the folders are shared from the host. These folders are not shared mounts (i doubt the fuse.osxfs or whatever they use) support that, or they simply don't care.

Docker for Mac now embeds its own kubelet, I believe they will have to deal with this once they hit this problem. In OpenShift we spotted this sooner as we just recently rebased to kube 1.10...

**Comment by mfojtik** [comment]
Also to add context, OpenShift 3.10 "cluster up" will run static pods and we need hostPath to add config into these pods (static pods will run api server, so you can't use configmaps/etc...). If you're on Docker for Mac, cluster up will run the container with (our) kubelet in Docker VM. This kubelet is based on 1.10, where the Docker kubelet versioning or origin is a mystery for me ;-)
